conda/core/solve.py

```diff
@@ -1029,6 +1039,251 @@ def _prepare(self, prepared_specs):
+                if not pkg.name.startswith("__"):  # skip virtual packages
+                    specs_to_add.append(MatchSpec(pkg.name))
+        # C. Implicitly requested via `install|update --update-deps`
+        elif context.update_modifier == UpdateModifier.UPDATE_DEPS:
+            final_specs = specs_to_add
+            for spec in specs_to_add:
+                prec = next(p for p in state["installed_pkgs"] if p.name == spec.name)
+                for dep in prec.depends:
+                    ms = MatchSpec(dep)
+                    if ms.name != "python":
+                        final_specs.append(MatchSpec(ms.name))
+            specs_to_add = list(set(final_specs))
+        return [s.conda_build_form() for s in specs_to_add]
+
+    def _pin_python(self, state):
+        """
+        If Python is already present in the environment and the user didn't
+        include it in the requested packages, make sure it remains pinned to
+        the minor version.
+        """
+        installed_names = [rec.name for rec in state["installed_pkgs"]]
+        is_installed = "python" in installed_names
+        is_explicitly_requested = any(s == "python" for s in self.specs_to_add_names)
+        is_requested_in_update_all = context.update_modifier == UpdateModifier.UPDATE_ALL
+        if is_installed and (is_requested_in_update_all or not is_explicitly_requested):
+            version = state["installed_pkgs"][installed_names.index("python")].version
+            major_minor_version = ".".join(version.split(".")[:2])
+            return f"python {major_minor_version}.*"
+
+    def _pinned_specs(self):
+        pinned_specs = get_pinned_specs(context.target_prefix)
+        if pinned_specs:
+            conda_prefix_data = PrefixData(context.target_prefix)
+        pin_these_specs = []
+        for s in pinned_specs:
+            x = conda_prefix_data.query(s.name)
+            if x:
+                for el in x:
+                    if not s.match(el):
+                        raise SpecsConfigurationConflictError([x], [el], self.prefix)
+            pin_these_specs.append(s.conda_build_form())
+        return pin_these_specs
+
+    def _run_solver(self, state):
+        solver = state["solver"]
+        solved = solver.solve()
+        # 4. Report problems if any
+        if not solved:
+            # it would be better if we could pass a graph object or something
+            # that the exception can actually format if needed
+            raise CondaError(solver.problems_to_str(), caused_by=self.__class__.__name__)
+
+    def _build_transaction(self, state):
+        from mamba import mamba_api as api
+        from mamba.utils import to_txn
+
+        transaction = api.Transaction(
+            state["solver"],
+            api.MultiPackageCache(context.pkgs_dirs),
+            PackageCacheData.first_writable().pkgs_dir,
+        )
+        (names_to_add, names_to_remove), to_link, to_unlink = transaction.to_conda()
```

Comment:
Mind elaborating what `to_conda` does? This seems like something we want to have implemented locally to be able to adopt to changes inside Conda easier.

Reply:
This is going to be tricky because it comes from the C++ API directly :/ I have simplified how we relied on `mamba.utils.to_txn` in #10893 and #10894.